src/plugins/azgenerator/templates/TemplateBase.ts

```diff
@@ -0,0 +1,44 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+import { Host } from "@azure-tools/autorest-extension-base";
+import { EOL } from "os";
+import { isNullOrUndefined } from "util";
+import { GenerationMode } from "../../models";
+import { CodeModelAz } from "../CodeModelAz";
+
+export abstract class TemplateBase {
+    private host: Host;
+    protected model: CodeModelAz;
+    protected isDebugMode: boolean;
+    public fullPath: string;
+
+    constructor(host: Host, model: CodeModelAz, isDebugMode: boolean) {
+        this.host = host;
+        this.model = model;
+        this.isDebugMode = isDebugMode;
+        this.fullPath = "";
+    }
+
+    public async work(): Promise<void> {
+        // Determine full or incremental
+        let generatedResults: string[];
+        if (this.model.Extension_GenerationMode == GenerationMode.Full) {
+            generatedResults = this.fullGeneration();
+        }
+        else {
+            let base: string[] = (await this.host.ReadFile(this.fullPath)).split(EOL);
+            generatedResults = this.incrementalGeneration(base);
+        }
+
+        if (!isNullOrUndefined(generatedResults)) {
+            this.host.WriteFile(this.fullPath, generatedResults.join(EOL));
+        }
+    }
```

Comment:
Don't see the neccessity of `if (this.model.Extension_GenerationMode == GenerationMode.Full) ` here. This work() are only called in `src/plugins/azgenerator/AzExtensionIncrementalGenerator.ts` which already implicates that it is for incremetal. 
If you are going to make it work under both fullGenerator and incrementalGenerator, you need to put it at azgenerator.ts before `await AzGeneratorFactory.createAzGenerator`, but in that way why split fullGenerator and IncrementalGenerator.

Reply:
The target for work here is to replace current template in full/incremental generator. It's a little bit dup... But full/incremental generator will show which file will be changed more clearly in code level. 

---
In reply to: [472726758](https://github.com/Azure/autorest.az/pull/519#discussion_r472726758) [](ancestors = 472726758)